nexus/src/app/sagas/snapshot_create.rs

```diff
@@ -0,0 +1,765 @@
+            .map_err(ActionError::action_failed)?;
+
+        warn!(log, "crucible region {:?}", crucible_region);
+
+        let crucible_snapshot = client
+            .region_get_snapshot(
+                &RegionId(region.id().to_string()),
+                &snapshot_id.to_string(),
+            )
+            .await
+            .map_err(|e| e.to_string())
+            .map_err(ActionError::action_failed)?;
+
+        warn!(log, "crucible snapshot {:?}", crucible_snapshot);
+
+        // Start the snapshot running
+        let crucible_running_snapshot = client
+            .region_run_snapshot(
+                &RegionId(region.id().to_string()),
+                &snapshot_id.to_string(),
+            )
+            .await
+            .map_err(|e| e.to_string())
+            .map_err(ActionError::action_failed)?;
+
+        warn!(log, "crucible running snapshot {:?}", crucible_running_snapshot);
+
+        // Map from the region to the snapshot
+        let region_addr = format!(
+            "{}",
+            dataset.address_with_port(crucible_region.port_number)
+        );
+        let snapshot_addr = format!(
+            "{}",
+            dataset.address_with_port(crucible_running_snapshot.port_number)
+        );
+        warn!(log, "map {} to {}", region_addr, snapshot_addr);
+        map.insert(region_addr, snapshot_addr);
+    }
+
+    Ok(map)
+}
+
+async fn ssc_start_running_snapshot_undo(
+    sagactx: ActionContext<SagaSnapshotCreate>,
+) -> Result<(), anyhow::Error> {
+    let log = sagactx.user_data().log();
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params();
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+
+    let snapshot_id = sagactx.lookup::<Uuid>("snapshot_id")?;
+
+    let (.., disk) = LookupPath::new(&opctx, &osagactx.datastore())
+        .project_id(params.project_id)
+        .disk_name(&params.create_params.disk.clone().into())
+        .fetch()
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    // Delete any running snapshots created by this saga
+    let datasets_and_regions = osagactx
+        .datastore()
+        .get_allocated_regions(disk.volume_id)
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    for (dataset, region) in datasets_and_regions {
+        // Create a Crucible agent client
+        let url = format!("http://{}", dataset.address());
+        let client = CrucibleAgentClient::new(&url);
+
+        // Delete running snapshot
+        warn!(
+            log,
+            "deleting running snapshot {} {} {}",
+            url,
+            region.id(),
+            snapshot_id
+        );
+        client
+            .region_delete_running_snapshot(
+                &RegionId(region.id().to_string()),
+                &snapshot_id.to_string(),
+            )
+            .await
+            .map_err(|e| e.to_string())
+            .map_err(ActionError::action_failed)?;
```

Comment:
Looking at the simulated crucible agent, this doesn't appear to do what we want. If the regions were partially allocated, or if this undo action was executed twice, it seems like this request would be sent to endpoints which previously executed the `DELETE` operation, and would see an error response.

If I'm reading `sled-agent/src/sim/http_entrypoints_storage.rs` correctly, this would return an error on the second invocation, or when an invocation was made to a region that did not get allocated in the "forward" action:

```rust
#[endpoint {
    method = DELETE,
    path = "/crucible/0/regions/{id}/snapshots/{name}/run",
}]
async fn region_delete_running_snapshot(
    rc: Arc<RequestContext<Arc<CrucibleData>>>,
    path: TypedPath<RunSnapshotPath>,
) -> Result<HttpResponseDeleted, HttpError> {
    let p = path.into_inner();
    let crucible = rc.context();

    if crucible.get(p.id.clone()).await.is_none() {
        return Err(HttpError::for_not_found(
            None,
            "Region not found".to_string(),
        ));
    }
...
```

Reply:
This specific error is for if the region does not exist, not the snapshot. The `delete_snapshot` call below is idempotent.

However, `create_running_snapshot` and `delete_running_snapshot` were both not idempotent, so that was fixed in 794eea13.